tools/diagnose/src/cmd/walk.rs

```diff
@@ -0,0 +1,176 @@
+use std::path::PathBuf;
+
+use clap::{Parser, ValueEnum};
+use regex::{Regex, RegexSet};
+use stable_eyre::{eyre::Context, Result};
+use strum::Display;
+use tracing::{debug, info, trace, warn};
+use walkdir::{DirEntry, WalkDir};
+
+/// Arguments used by the "walk" command.
+#[derive(Debug, Clone, Parser)]
+#[command(version, about)]
+pub struct WalkArgs {
+    /// Don't walk paths that match any of the provided regular expressions.
+    ///
+    /// File paths that consist of invalid UTF-8 data are lossily converted
+    /// using the unicode replacement character (U+FFFD).
+    ///
+    /// Note: this includes subdirectories;
+    /// FOSSA CLI filtering is much more complicated so is not a 1:1 comparison.
+    /// For more information, see: https://github.com/fossas/fossa-cli/blob/master/docs/contributing/filtering.md
+    #[clap(long = "filter")]
+    filters: Vec<Regex>,
+
+    /// Control whether symbolic links (symlinks) are followed.
+    ///
+    /// FOSSA CLI follows symlinks by default, so this is defaulted to following symlinks.
+    /// Broken or looping symlinks are warned and then ignored.
+    #[clap(long, default_value_t = SymlinkStrategy::Follow)]
+    symlinks: SymlinkStrategy,
+}
+
+/// The strategy for how to handle symbolic links while walking the file system.
+#[derive(Debug, Clone, Copy, ValueEnum, Display)]
+pub enum SymlinkStrategy {
+    /// Symbolic links are followed.
+    #[strum(serialize = "follow")]
+    Follow,
+
+    /// Symbolic links are ignored.
+    #[strum(serialize = "skip")]
+    Skip,
+}
+
+#[tracing::instrument(skip(args), fields(root_device_id))]
+pub fn main(root: PathBuf, args: &WalkArgs) -> Result<()> {
+    info!("inspecting root: {}", root.display());
+    trace!("inspecting root: {}", root.display());
+    debug!("walk options: {args:?}");
+
+    let root_device_id = util::device_num(&root).wrap_err("get device ID")?;
+    tracing::Span::current().record("root_device_id", root_device_id);
+    trace!("device_id: {root_device_id}");
```

Comment:
If we find that the root device is different than one of the walked files, should we run something like `mount -v` in order to get filesystem info? I'm not sure what the windows equivalent is.

If you think it's safe, just unconditionally capturing the output of that command for cross-reference may also work. I don't think there should be anything sensitive in it.

Reply:
Hmm, I'd rather leave this as is for now, since we aren't certain we need this information, but agreed we can add that if it is something we run into that we want for debugging.